src/test/test_circuitpadding.c

```diff
@@ -0,0 +1,975 @@
+  tt_int_op(circpad_histogram_bin_to_usec(relay_side->padding_info[0], 0),
+            OP_EQ,
+            relay_side->padding_info[0]->rtt_estimate_us+
+            circpad_machine_current_state(
+             relay_side->padding_info[0])->start_usec);
+
+  /* Test 2: Termination of RTT measurement (from the previous test) */
+  tt_int_op(relay_side->padding_info[0]->stop_rtt_update, OP_EQ, 1);
+  rtt_estimate = relay_side->padding_info[0]->rtt_estimate_us;
+
+  circpad_cell_event_nonpadding_received((circuit_t*)relay_side);
+  timers_advance_and_run(4);
+  circpad_cell_event_nonpadding_sent((circuit_t*)relay_side);
+
+  tt_int_op(relay_side->padding_info[0]->rtt_estimate_us, OP_EQ, rtt_estimate);
+  tt_int_op(relay_side->padding_info[0]->last_received_time_us, OP_EQ, 0);
+  tt_int_op(relay_side->padding_info[0]->stop_rtt_update, OP_EQ, 1);
+  tt_int_op(circpad_histogram_bin_to_usec(relay_side->padding_info[0], 0),
+            OP_EQ,
+            relay_side->padding_info[0]->rtt_estimate_us+
+            circpad_machine_current_state(
+             relay_side->padding_info[0])->start_usec);
+
+  /* Test 3: Make sure client side machine properly ignores RTT */
+  circpad_cell_event_nonpadding_received((circuit_t*)client_side);
+  tt_int_op(client_side->padding_info[0]->last_received_time_us, OP_EQ, 0);
+
+  timers_advance_and_run(20);
+  circpad_cell_event_nonpadding_sent((circuit_t*)client_side);
+  tt_int_op(client_side->padding_info[0]->last_received_time_us, OP_EQ, 0);
+
+  tt_int_op(client_side->padding_info[0]->rtt_estimate_us, OP_EQ, 0);
+  tt_int_op(circpad_histogram_bin_to_usec(client_side->padding_info[0], 0),
+            OP_NE, client_side->padding_info[0]->rtt_estimate_us);
+  tt_int_op(circpad_histogram_bin_to_usec(client_side->padding_info[0], 0),
+            OP_EQ,
+            circpad_machine_current_state(
+                client_side->padding_info[0])->start_usec);
+ done:
+  free_fake_orcirc(relay_side);
+  circuitmux_detach_all_circuits(dummy_channel.cmux, NULL);
+  circuitmux_free(dummy_channel.cmux);
+  timers_shutdown();
+  monotime_disable_test_mocking();
+  UNMOCK(circuit_package_relay_cell);
+  UNMOCK(circuitmux_attach_circuit);
+
+  return;
+}
+
+static circpad_machine_t circ_client_machine;
+void
+circpad_circ_token_machine_setup(circuit_t *on_circ)
+{
+  on_circ->padding_machine[0] = &circ_client_machine;
+  on_circ->padding_info[0] = circpad_circuit_machineinfo_new(on_circ, 0);
+
+  circ_client_machine.start.transition_events[CIRCPAD_STATE_BURST] =
+    CIRCPAD_EVENT_NONPADDING_RECV;
+
+  circ_client_machine.burst.transition_events[CIRCPAD_STATE_BURST] =
+    CIRCPAD_EVENT_PADDING_RECV |
+    CIRCPAD_EVENT_NONPADDING_RECV;
+
+  circ_client_machine.burst.transition_cancel_events =
+    CIRCPAD_EVENT_NONPADDING_SENT;
+
+  // FIXME: Is this what we want?
+  circ_client_machine.burst.token_removal = CIRCPAD_TOKEN_REMOVAL_HIGHER;
+
+  // FIXME: Tune this histogram
```

Comment:
So I've been writing unittests. I wanted to run a unittest that tests more token removal functions. I started off by copying the setup logic of `test_circuitpadding_tokens()` but then I wanted to verify the width of the various bins of the histogram below. I know that the first bin should be of width 0, and after that the bins should double in size until `range_usec`.

I tried to verify this assumption by writing code like this:
```
  /* Test first bin width */
  circpad_delay_t d1 = circpad_histogram_bin_to_usec(mi, 0);
  circpad_delay_t d2 = circpad_histogram_bin_to_usec(mi, 1)-1;
  uint32_t first_bin_width = d2-d1;
  printf("first_bin_width: %u\n", first_bin_width);

  d1 = circpad_histogram_bin_to_usec(mi, 1);
  d2 = circpad_histogram_bin_to_usec(mi, 2)-1;
  uint32_t second_bin_width = d2-d1;
  printf("second_bin_width: %u\n", second_bin_width);

  d1 = circpad_histogram_bin_to_usec(mi, 2);
  d2 = circpad_histogram_bin_to_usec(mi, 3)-1;
  uint32_t third_bin_width = d2-d1;
  printf("third_bin_width: %u\n", third_bin_width);

  d1 = circpad_histogram_bin_to_usec(mi, 3);
  d2 = circpad_histogram_bin_to_usec(mi, 4)-1;
  uint32_t fourth_bin_width = d2-d1;
  printf("fourth_bin_width: %u\n", fourth_bin_width);
```

And here are the outputs:
```
first_bin_width: 0
second_bin_width: 249998
third_bin_width: 249999
fourth_bin_width: 4294466794
```

I think it's normal that the 4th bin comes out to UINT_MAX because the `bin_to_usec()` function returns infinite delay if `bin >= CIRCPAD_INFINITY_BIN(state)`, but I don't understand why the other two bins don't come out to exponential size.

Maybe what I'm doing with subtracting -1 from bin_to_usec is not right, but I got that from the documentation of the function:
```
 * Calculate the lower bound of a histogram bin. The upper bound
 * is obtained by calling this function with bin+1, and subtracting 1.
```

Reply:
Err can we try more than one two datapoints here? The actual property I was going for was expoentially spaces bin cutoffs. Saying that the bins are doubling is perhaps poor phrasing, but it should actually happen after the third bin (in a histogram with enough bins).

Reply:
ACK. I'm on the process of making this test better and more robust. Here are the answers for a histogram with:
```
  burst_state->histogram_len = 15;
  burst_state->start_usec = 0;
  burst_state->range_usec = 100000;
```

Here we go:
```
 0 bin_width: 0
1 bin_width: 22
2 bin_width: 23
3 bin_width: 48
4 bin_width: 97
5 bin_width: 194
6 bin_width: 390
7 bin_width: 780
8 bin_width: 1562
9 bin_width: 3124
10 bin_width: 6249
11 bin_width: 12499
12 bin_width: 24999
13 bin_width: 4294917294
14 bin_width: 4294967295
```